### PR TITLE
Sentiment analysis using Pipelines

### PR DESCRIPTION
**Pull request by payoto**
This PR is for Graphcore internal review of the notebook, the notebook will eventually be committed without the outputs. To facilitate review a python file has been generated please add comments to `sentiment_analysis.py` I will then feed that back into the notebook.

You can see the rendered notebook with outputs at: https://github.com/graphcore/optimum-graphcore-fork/blob/b77f3e17ab8cf6f5ea8b1f17fe179ec9ccfd37b2/notebooks/sentiment_analysis.ipynb

Things that need doing:

- [x] Make it work with our caching mechanism in Paperspace.
- [x] Strip the outputs
- [x] remove the python file
- [x] Make guide on managing IPU resources from Notebook
  - [x] Add links

### CODE REVIEW

File: notebooks/sentiment_analysis.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+"""
+## Using transformers pipelines on the IPU
+
+The simplest way to use a model on the IPU is to use the `pipeline` function. It provides a set of models which have been validated to work on a given task, to get started choose the task and call the `pipeline` function:
+"""
+from optimum.graphcore import pipelines
+sentiment_pipeline = pipelines.pipeline("sentiment-analysis")
+
+"""
+Because no model was supplied the pipeline defaulted to `distilbert-base-uncased-finetuned-sst-2-english`. We are helpfully warned that we should provide the `max_length` argument if we were to put this model in production, but while we are experimenting we will leave it as is.
```
Comment (ncouro-gc) [inline notebooks/sentiment_analysis.py:37]: nit: Maybe remove this line

File: notebooks/sentiment_analysis.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+Now it's time to test our first prompts. Let's start with some very easy to classify text:
+"""
+simple_test_data = ["I love you", "I hate you"]
+sentiment_pipeline(simple_test_data)
+
+"""
+Reassuringly, the model got it right! And with a high degree of confidence, more than 99.9% in both cases.
+
+You'll have noticed that this was a bit slow, it took several seconds to provide the answer. This is because of the computing architecture of the IPU, during the first execution of a pipeline, the model needs to be compiled and loaded into the IPU. While this takes some time, these overheads only the first execution of a pipeline!
```
Comment (ncouro-gc) [inline notebooks/sentiment_analysis.py:47]: I would rephrase that - no need to say "noticed". Just something like "Compilation of the model happens at the first call, subsequent calls will be faster"

File: notebooks/sentiment_analysis.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+sentiment_pipeline(simple_test_data)
+
+"""
+Now that was much faster! We can use the `%%timeit` cell magic to check how fast:
+"""
+% % timeit
+sentiment_pipeline(simple_test_data)
+
+"""
+It takes on the order of ~1ms per prompt, this is fast!
```
Comment (ncouro-gc) [inline notebooks/sentiment_analysis.py:60]: Maybe scrap "this is fast" until we know we're definitely faster than a A100?

File: notebooks/sentiment_analysis.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+"""
+"""
+### Customising the IPU resources
```
Comment (ncouro-gc) [inline notebooks/sentiment_analysis.py:63]: I'm not sure this section is needed.

File: notebooks/sentiment_analysis.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+emotion_model([
+    "How are you today?",
+    "Don't make me go out, it's too cold!",
+    "What is happening, I don't understand",
+    "Where did you come from?",
+])
+
+
+"""
+## Managing resources
```
Comment (ncouro-gc) [inline notebooks/sentiment_analysis.py:258]: I'd consider moving this in a separate notebook ("managing IPU resources from notebooks")

File: notebooks/sentiment_analysis.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+While we see a slight drop in the confidence of the model, but it does not feel sufficient to reflect message.
+
+The imprecise classification of these prompts would affect any downstream task: if we were trying to derive some insights from the model on customer satisfaction we may have an overly optimistic view of performance and derive the wrong conclusions.
+
+To resolve this issue we need to try more models, which accommodate finer grained classification.
+"""
+"""
+## Trying more models
+
+[The blog post]() we are following suggests a number of other models, lets try them all to see if they perform better on our ambiguous prompts!
```
Comment (ncouro-gc) [inline notebooks/sentiment_analysis.py:108]: add link

File: notebooks/sentiment_analysis.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+Integration of the Graphcore Intelligence processing unit (IPU) and the HuggingFace transformer library means that it only takes a few lines of code to perform complex tasks which require deep learning.
+
+In this notebook we perform **sentiment analysis**: we use natural language processing models to classify text prompts. 
+We follow [this blog post by Federico Pascual](https://huggingface.co/blog/sentiment-analysis-python) and test 5 different models available on HuggingFace Hub to highlight different model properties of the models that can be leveraged for downstream tasks.
+
+The ease of use of the `pipeline` interface lets us quickly experiment with the pre-trained models and identify which one will work best.
+This simple interface means that it is extremely easy to access the fast inference performance of the IPU on your application.
+
+![Widget inference on a text classification task](images/text_classification.png)
```
[the remark this answers is not in the document]
Comment (payoto) [reply notebooks/sentiment_analysis.py:12]: Yep

File: notebooks/sentiment_analysis.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+%load_ext autoreload
+%autoreload 2
+
+import os
+os.environ["POPTORCH_LOG_LEVEL"] = "ERR"
+
+"""
+## Using transformers pipelines on the IPU
+
+The simplest way to use a model on the IPU is to use the `pipeline` function. It provides a set of models which have been validated to work on a given task, to get started choose the task and call the `pipeline` function:
```
[the remark this answers is not in the document]
Comment (payoto) [reply notebooks/sentiment_analysis.py:31]: Well, there wasn't so I'm adding `pipelines.list_tasks()`, I'll move it up.

File: notebooks/sentiment_analysis.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+multilingual_model(ambiguous_in_french)
+
+
+"""
+The model gives very similar results in both languages!
+
+Now let us revisit our first model, could it also work with this multi-lingual input?
+"""
+print(sentiment_pipeline(ambiguous_prompts))
+print(sentiment_pipeline(ambiguous_in_french))
```
Comment (anjleeg-gcai) [inline notebooks/sentiment_analysis.py:222]: Can you rename `sentiment_pipeline` at the start to `sentiment_model` to fit in with the same structure as the other models used in this tutorial?

File: notebooks/sentiment_analysis.py (hunk 2 above)
Comment (payoto) [reply notebooks/sentiment_analysis.py:47]: yeah that was unnecessary I went with:

> The first call to the pipeline was a bit slow, it took several seconds to provide the answer. This behaviour is due to compilation of the model which happens on the first call.
On subsequent prompts it is much faster:

File: notebooks/sentiment_analysis.py (hunk 3 above)
Comment (payoto) [reply notebooks/sentiment_analysis.py:60]: https://developer.nvidia.com/deep-learning-performance-training-inference

Should I rephrase to:

> This is [_fast_](https://developer.nvidia.com/deep-learning-performance-training-inference)!

😉

Comment (payoto) [reply notebooks/sentiment_analysis.py:60]: NVidia A100 40GB MLPerf 2.1 inference at BS1 with triton is 4ms

File: notebooks/sentiment_analysis.py (hunk 4 above)
Comment (payoto) [reply notebooks/sentiment_analysis.py:63]: Well I hate it, but I need the `inference_config` to be set otherwise nothing works the alternative is to change defaults in the library. Which I'd like to do after the release but seems risky before. 

Given that I prefer addressing the arguments front and centre, I feel it is beneficial to drip feed a bit of IPU information in the tutorial so that if they then read tutorials that are more IPU specific it's not a total surprise that the integration is not perfect.

File: notebooks/sentiment_analysis.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+    *"I'm a little tired, I didn't sleep well, but I hope it gets better".split(","),
+])
+out
+
+"""
+Here each parts of the sentence is correctly classified, how you choose to process those sentence parts will depend on what you need to use the results of the sentiment analysis for.
+
+We can do small changes to the prompt to get an intuition of how the model responds to changes in grammar. Below the last part of the ambiguous sentence is changed to be more optimistic:
+"""
+print(f"Previous score: {out[-1]}")
```
[the remark this answers is not in the document]
Comment (payoto) [reply notebooks/sentiment_analysis.py:140]: Good catch!

File: notebooks/sentiment_analysis.py (hunk 5 above)
Comment (payoto) [reply notebooks/sentiment_analysis.py:258]: Good shout, like that we can point to it from different notebooks.

File: notebooks/sentiment_analysis.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+"""
+"""
+### A model finetuned on tweets
+
+The next model discussed in the blog post is the [`cardiffnlp/twitter-roberta-base-sentiment-latest`](https://huggingface.co/cardiffnlp/twitter-roberta-base-sentiment-latest) model, it is a RoBERTa-Base which was trained on 124M tweets collected between 2018 and 2021. This data makes the model much more recent than the previous pre-trained checkpoint.
+
+As before this model is trivial to load through the pipeline API:
+"""
+from pprint import pprint
+newer_model = pipelines.pipeline(
```
[the remark this answers is not in the document]
Comment (payoto) [reply notebooks/sentiment_analysis.py:155]: Great suggestion!

File: notebooks/sentiment_analysis.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+# on managing resources at the end of the document.
+tweet_model.model.detachFromDevice()
+newer_model.model.detachFromDevice()
+
+"""
+### Multi-lingual model
+
+The next model has an interesting feature: it is multi-lingual. It was trained on a dataset of English, Dutch, German, Spanish, Italian and French text, it can be prompted in any of these languages and should correctly classify the text inputs.
+
+The model is `nlptown/bert-base-multilingual-uncased-sentiment` a BERT checkpoint fine-tuned on ~700k reviews in 6 languages and predicts between 1 and 5 stars for each prompt.
```
[the remark this answers is not in the document]
Comment (payoto) [reply notebooks/sentiment_analysis.py:187]: Added:

> 1 start indicates a very negative sentiment, while 5 starts corresponds to a very positive text.

Does this hit the mark?

File: notebooks/sentiment_analysis.py (hunk 9 above)
Comment (payoto) [reply notebooks/sentiment_analysis.py:222]: Renamed all to `pipeline`

File: notebooks/sentiment_analysis.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (anjleeg-gcai) [reply notebooks/sentiment_analysis.py:187]: Yeah sounds good!

**Comment by payoto** [comment]
Addressed review comments, now going to make a quick notebook on managing resources from Notebooks.

File: notebooks/sentiment_analysis.ipynb [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+    "In this case we see the multi-lingual model correctly predict strongly positive and negative labels, while the other model predicts a positive message (correct) and a neutral (expected LABEL_0)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Other models\n",
+    "\n",
+    "Models can be finetuned to extract different classes from text. The `bhadresh-savani/distilbert-base-uncased-emotion` checkpoint is a DistilBERT checkpoint tuned to identify the emotion associated with a prompt:"
```
Comment (JamesRandom) [inline notebooks/sentiment_analysis.ipynb:539]: The wording "The `bhadresh-savani/distilbert-base-uncased-emotion` checkpoint is a DistilBERT checkpoint" is a bit clunky. Can we change one "checkpoint" to "model"? Or remove one?

**Comment by payoto** [comment]
Thanks all for the reviews and @JamesRandom for the updates!
